forte/processors/data_augment/algorithms/skeleton_op.py

```diff
@@ -0,0 +1,765 @@
+        if last_span_ind >= 0:
+            if is_inclusive:
+                if is_begin:
+                    # When inclusive, move the begin index
+                    # to the left to include the inserted span.
+                    if (
+                        last_span_ind > 0
+                        and old_spans[last_span_ind - 1].begin == index
+                    ):
+                        # Old spans: [0, 1], [1, 1], [1, 3]
+                        # Target index: 1
+                        # Change last_span_index from 2 to 1
+                        # to include the [1, 1] span.
+                        last_span_ind -= 1
+                    else:
+                        # Old spans: [0, 1], [1, 1], [2, 3]
+                        # Target index: 1
+                        # last_span_index: 1
+                        # No need to change.
+                        pass
+
+            else:
+                if not is_begin:
+                    # When exclusive, move the end index
+                    # to the left to exclude the inserted span.
+                    if (
+                        last_span_ind > 0
+                        and old_spans[last_span_ind - 1].begin == index
+                    ):
+                        # Old spans: [0, 1], [1, 1], [1, 3]
+                        # Target index: 1
+                        # Change last_span_index from 2 to 0
+                        # to exclude the [1, 1] span.
+                        last_span_ind -= 2
+                    elif (
+                        old_spans[last_span_ind].begin == index
+                        and old_spans[last_span_ind].end == index
+                    ):
+                        # Old spans: [0, 1], [1, 1], [2, 3]
+                        # Target index: 1
+                        # Change last_span_index from 1 to 0
+                        # to exclude the [1, 1] span.
+                        last_span_ind -= 1
+
+        if last_span_ind < 0:
+            # There is no replacement before this index.
+            return index
+        # Find the nearest anchor point on the left of current index.
+        # Start from the span's begin index.
+        delta_index: int = (
+            new_spans[last_span_ind].begin - old_spans[last_span_ind].begin
+        )
+
+        if (
+            old_spans[last_span_ind].begin == old_spans[last_span_ind].end
+            and old_spans[last_span_ind].begin == index
+            and is_begin
+            and is_inclusive
+        ):
+            return index + delta_index
+
+        if old_spans[last_span_ind].end <= index:
+            # Use the span's end index as anchor, if possible.
+            delta_index = (
+                new_spans[last_span_ind].end - old_spans[last_span_ind].end
+            )
+        return index + delta_index
+
+    def _apply_augmentations(
+        self,
+        data_pack: DataPack,
+    ) -> DataPack:
```

Comment:
Simiarly, is there anything new here or mostly copied? I will skip reviewing this if they are copied.

Reply:
There is nothing new here